jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/BaseKDisjointShortestPathsAlgorithm.java

```diff
@@ -0,0 +1,260 @@
+            .stream().flatMap(List::stream).filter(e -> !this.overlappingEdges.contains(e))
+            .collect(Collectors.toSet());
+
+        for (E e : nonOverlappingEdges) {
+            V u = workingGraph.getEdgeSource(e);
+            if (u.equals(startVertex)) { // start of a new path
+                List<E> path = new ArrayList<>();
+                path.add(e);
+                paths.add(path);
+            } else { // some edge which is part of a path
+                if (!sourceToEdgeLookup.containsKey(u)) {
+                    sourceToEdgeLookup.put(u, new ArrayDeque<>());
+                }
+                sourceToEdgeLookup.get(u).add(e);
+            }
+        }
+
+        // Build the paths using the lookup table
+        for (List<E> path : paths) {
+            V v = workingGraph.getEdgeTarget(path.get(0));
+            while (!v.equals(endVertex)) {
+                E e = sourceToEdgeLookup.get(v).poll();
+                path.add(e);
+                v = workingGraph.getEdgeTarget(e);
+            }
+        }
+        
+        return paths
+            .stream().map(path -> createGraphPath(new ArrayList<>(path), startVertex, endVertex))
+            .collect(Collectors.toList());
+    }
+
+    /**
+     * Iterate over all paths to remove overlapping edges (i.e. those edges contained in more than 
+     * one path).
+     * Two edges are considered as overlapping in case both edges connect the same vertex pair, 
+     * disregarding direction.
+     * At the end of this method, each path contains unique edges but not necessarily connecting the
+     * start to end vertex.
+     * 
+     */
+    private void findOverlappingEdges()
+    {
+        Map<UnorderedPair<V, V>, Integer> edgeOccurrenceCount = new HashMap<>();
+        for (List<E> path : pathList) {
+            for (E e : path) {                
+                V v = this.workingGraph.getEdgeSource(e);
+                V u = this.workingGraph.getEdgeTarget(e);                
+                UnorderedPair<V, V> edgePair = new UnorderedPair<>(v, u);
+                
+                if (edgeOccurrenceCount.containsKey(edgePair)) {
+                    edgeOccurrenceCount.put(edgePair, 2);
+                } else {
+                    edgeOccurrenceCount.put(edgePair, 1);
+                }
+            }
+        }
+
+        this.overlappingEdges = pathList.stream().flatMap(List::stream).filter(
+            e -> edgeOccurrenceCount.get(new UnorderedPair<>(
+                this.workingGraph.getEdgeSource(e), 
+                this.workingGraph.getEdgeTarget(e))) > 1)
+            .collect(Collectors.toSet());
+    }
+
+    private GraphPath<V, E> createGraphPath(List<E> edgeList, V startVertex, V endVertex)
+    {
+        double weight = 0;
+        for (E edge : edgeList) {
+            weight += originalGraph.getEdgeWeight(edge);
+        }
+        return new GraphWalk<>(originalGraph, startVertex, endVertex, edgeList, weight);
+    }
+
+
+    /**

```

Comment:
(b) It's really preparing the graph for the next iteration, and, as you mentioned below, redundant in the last iteration. So, I think the other way around: it might confuse coming after calculate and not before. So, maybe the following would compromise between: First calculate call would be done out of the iteration. Then would come a sequence of prepare-calculate calls up to the k-th iteration. Javadoc would be updated.

(a) Done.